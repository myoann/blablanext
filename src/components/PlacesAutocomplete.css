.placesAutocomplete {
  position: relative;
}

.placesAutocomplete ul {
  position: absolute;
  top: 100%;
  left: 0;
  right: 0;
  z-index: 1000;
  margin: 0;
  padding: 0;
  list-style: none;
  background-color: #fff;
  border: 1px solid #ccc;
  border-top: 0;
  border-radius: 0 0 4px 4px;
  z-index: 5;
}

.placesAutocomplete li {
  padding: 10px;
  cursor: pointer;
  transition: background-color 0.2s;
}

.placesAutocomplete li:hover {
  background-color: #f0f0f0;
}

.placesAutocomplete li strong {
  font-weight: bold;
}

.placesAutocomplete li small {
  color: #888;
}
