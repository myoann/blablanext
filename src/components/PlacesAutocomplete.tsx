import React from "react";
import usePlacesAutocomplete, {
  getGeocode,
  getLatLng,
} from "use-places-autocomplete";

import { LatLng } from "@/types";

import "./PlacesAutocomplete.css";

type Props = {
  id: string;
  onCoordinates: (coords: LatLng) => void;
  placeholder: string;
};

const PlacesAutocomplete = ({ id, onCoordinates, placeholder }: Props) => {
  const {
    ready,
    value,
    suggestions: { status, data },
    setValue,
    clearSuggestions,
  } = usePlacesAutocomplete({
    debounce: 300,
    requestOptions: {
      componentRestrictions: {
        // Restrict to France and its neighbors
        country: ["fr", "be", "de", "es", "it", "gb", "ch", "lu", "mc", "ad"],
      },
    },
  });

  const handleInput = (e: React.ChangeEvent<HTMLInputElement>) => {
    setValue(e.target.value);
  };

  const handleSelect =
    ({ description }: { description: string }) =>
    () => {
      setValue(description, false);
      clearSuggestions();

      getGeocode({ address: description }).then((results) => {
        const { lat, lng } = getLatLng(results[0]);
        onCoordinates({ lat, lng });
      });
    };

  const renderSuggestions = () =>
    data.map((suggestion) => {
      const {
        place_id,
        terms,
        structured_formatting: { main_text },
      } = suggestion;

      const country = terms[terms.length - 1].value;

      return (
        <li key={place_id} onClick={handleSelect(suggestion)}>
          {main_text}, {country}
        </li>
      );
    });

  return (
    <div className="placesAutocomplete">
      <input
        id={id}
        value={value}
        onChange={handleInput}
        disabled={!ready}
        placeholder={placeholder}
      />
      {status === "OK" && <ul>{renderSuggestions()}</ul>}
    </div>
  );
};

export default PlacesAutocomplete;
