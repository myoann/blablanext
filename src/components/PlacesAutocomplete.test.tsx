import { render, screen } from "@testing-library/react";

import PlacesAutocomplete from "./PlacesAutocomplete";

const defaultProps = {
  id: "placesAutocomplete",
  placeholder: "Search for a place",
  onCoordinates: jest.fn(),
};

describe("PlacesAutocomplete", () => {
  it("matches the snapshot", () => {
    const { container } = render(<PlacesAutocomplete {...defaultProps} />);

    expect(container).toMatchSnapshot();
  });

  it("has an input with the placeholder passed as prop", () => {
    render(<PlacesAutocomplete {...defaultProps} />);

    const input = screen.getByPlaceholderText("Search for a place");
    expect(input).toBeInTheDocument();
  });
});
