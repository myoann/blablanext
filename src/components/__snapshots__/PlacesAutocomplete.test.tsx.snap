// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PlacesAutocomplete matches the snapshot 1`] = `
<div>
  <div
    class="placesAutocomplete"
  >
    <input
      id="placesAutocomplete"
      placeholder="Search for a place"
      value=""
    />
    <ul />
  </div>
</div>
`;
